### DISCUSSION TITLE
The entity logical name account is not valid.

### DISCUSSION THREAD

**Discussion by goncalomelo8** [jordimontana82/fake-xrm-easy#534] [Q&A] [answered]
Hello, I am new to Unit Tests and Microsoft CRM Plugin unit testing and I have been having this issue with some of my tests.

No matter what I try, it always fails in the same point with the same error message: "The entity logical name account is not valid."

If anyone could help me solve this or point me in the right direction, it would be great. 

Below is my code

**Plugin Interface**

```
public class PostOperationCreate : IPlugin
    {
        public void Execute(IServiceProvider serviceProvider)
        {
            var context = (IPluginExecutionContext)serviceProvider.GetService(typeof(IPluginExecutionContext));
            var serviceFactory = (IOrganizationServiceFactory)serviceProvider.GetService(typeof(IOrganizationServiceFactory));
            var service = serviceFactory.CreateOrganizationService(context.UserId);
            if (context.InputParameters.Contains("Target") && context.InputParameters["Target"] is Entity)
            {
                var entity = (Entity)context.InputParameters["Target"];
                if (entity.LogicalName == "account")
                    CreateRelatedContact.Execute(entity, service);
            }
        }
    }
```

**Business Logic**

```
public class CreateRelatedContact
    {
        public static void Execute(Entity contextEntity, IOrganizationService service)
        {
            EntityReference erPrimaryContact = contextEntity.GetAttributeValue<EntityReference>("primarycontactid");
            if (erPrimaryContact != null)
            {
                return;
            }

            Entity newContact = new Entity("contact");
            newContact["parentcustomerid"] = contextEntity.ToEntityReference();
            newContact["lastname"] = contextEntity.GetAttributeValue<string>("name");
            newContact["telephone1"] = contextEntity.GetAttributeValue<string>("telephone1");
            newContact["emailaddress1"] = contextEntity.GetAttributeValue<string>("emailaddress1");
            Guid newContactId = service.Create(newContact);

            contextEntity["primarycontactid"] = new EntityReference(newContact.LogicalName, newContactId);
            service.Update(contextEntity);
        }
    }
```

**Unit Test**

```
[TestClass]
    public class CreateRelatedContact
    {
        [TestMethod]
        public void Account_With_No_PrimaryContact_Created()
        {
            // Create the fake context. This allows us to setup a mocked version of Dynamics for use later in the test
            var context = new XrmFakedContext();

            // Generate a GUID which will be assigned to the fake record. It is done here to allow the record to be retrieved later during the test
            Guid accountGuid = Guid.NewGuid();

            // Create the entity record which will be the target
            var target = new Entity("account")
            {
                Id = accountGuid,
                Attributes =
                {
                    ["name"] = "Unit Test",
                    ["telephone1"] = "963256954",
                    ["emailaddress1"] = "unittest@gmail.com"
                }
            };

            // Execute the plugin.
            context.ExecutePluginWithTarget<PostOperationCreate>(target);

            // Retrieve the update account from the fake context
            Entity updatedAccount = context.GetOrganizationService().Retrieve("account", accountGuid, new ColumnSet("primarycontactid"));

            // Assert that the value meets our expected value
            Assert.IsNotNull(updatedAccount.GetAttributeValue<EntityReference>("primarycontactid"));
        }
    }
```

**Comment by BetimBeja** [comment, marked as answer]
Hello @goncalomelo8,
in your unit test you are not actually creating the account, you are just testing your plugin with the default context which is PreOperation of Create. Also I guess your test is failing in the Update account request in your CreateRelatedContact class... because the account is not currently in the context. Adding it to the context and testing the plugin should make your test pass, even though, I don't think is the correct way to do your test.

**Reply by jordimontana82** [reply]
@goncalomelo8 Since you're using late bound, the error is thrown [because you're trying to retrieve / update an account  but there are no account records in the in memory database](https://github.com/jordimontana82/fake-xrm-easy/blob/5531248d7e9dc2a112b17735e0bf5939e974d342/FakeXrmEasy.Shared/XrmFakedContext.Crud.cs#L34). You need to either create the account before the plugin is executed, or else call context.Initialize(account) so the context (a.k.a in memory database) will know about it when retrieving it.

**Reply by goncalomelo8** [reply]
Yes, I did that and it fixed my problem, thanks @BetimBeja and @jordimontana82!

@BetimBeja why do you think its not the correct way to do the test and what do you advise?

**Reply by BetimBeja** [reply]
I think it is also important to test the created contact (if it has the correct fields populated) and you can get the necessary objects from the context by using `context.CreateQuery("contact")`.

**Comment by bwmodular** [comment]
Hi @goncalomelo8, I think the problem is that you haven't initialised the context so it doesn't 'know' what an 'account' is.  Take a look at this issue #406 - this will show you how to initialise the context so that the record you are retrieving in your test will exist and the context will know what an account is.

**Reply by goncalomelo8** [reply]
Yes, that was exactly the problem, thanks a lot @bwmodular!

**Comment by bwmodular** [comment]
@BetimBeja beat me to it!
I think he's right - if you create the contact as part of the PreOperation of the Create message, then you could add the primarycontactid to the target as part of the plugin, and not have to perform another update on the account just after creating it.

**Reply by goncalomelo8** [reply]
I am doing this on PostUpdate because I need the account reference to fill the "parentcustomerid" on the Contact that I am creating.
